hexa/datasets/models.py

```diff
@@ -0,0 +1,219 @@
+        if not principal.has_perm("datasets.share_dataset", self):
+            raise PermissionDenied
+
+        return DatasetShare.objects.create(
+            created_by=principal,
+            dataset=self,
+            workspace=workspace,
+        )
+
+
+class DatasetVersionQuerySet(BaseQuerySet):
+    def filter_for_user(self, user: typing.Union[AnonymousUser, User]):
+        # TODO: It should also check workspace where it's added
+        return self._filter_for_user_and_query_object(
+            user,
+            models.Q(dataset__workspace__members=user)
+            | models.Q(dataset__shares__workspace__members=user),
+            return_all_if_superuser=False,
+        )
+
+
+class DatasetVersionManager(models.Manager):
+    def create_if_has_perm(
+        self, principal: User, dataset: Dataset, *, name: str, description: str
+    ):
+        if not principal.has_perm("datasets.create_dataset_version", dataset):
+            raise PermissionDenied
+
+        version = self.create(
+            name=name,
+            dataset=dataset,
+            created_by=principal,
+            description=description,
+        )
+
+        create_bucket_folder(settings.DATASETS_BUCKET, f"{dataset.id}/{version.id}")
+
+        return version
+
+
+class DatasetVersion(Base):
+    dataset = models.ForeignKey(
+        Dataset,
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        related_name="versions",
+    )
+    name = models.TextField(null=False, blank=False)
+    description = models.TextField(blank=True, null=True)
+    created_by = models.ForeignKey(User, null=True, on_delete=models.SET_NULL)
+
+    objects = DatasetVersionManager.from_queryset(DatasetVersionQuerySet)()
+
+    class Meta:
+        ordering = ["-created_at"]
+
+    def delete_if_has_perm(self, *, principal: User):
+        if not principal.has_perm("datasets.delete_dataset_version", self):
+            raise PermissionDenied
+        self.delete()
+
+
+class DatasetVersionFileQuerySet(BaseQuerySet):
+    def filter_for_user(self, user: typing.Union[AnonymousUser, User]):
+        return self._filter_for_user_and_query_object(
+            user,
+            models.Q(
+                dataset_version__dataset__in=Dataset.objects.filter_for_user(user)
+            ),
+            return_all_if_superuser=False,
+        )
+
+
+class DatasetVersionFile(Base):
+    uri = models.TextField(null=False, blank=False, unique=True)
+    content_type = models.TextField(null=False, blank=False)
+    created_by = models.ForeignKey(User, null=True, on_delete=models.SET_NULL)
+    dataset_version = models.ForeignKey(
+        DatasetVersion,
+        null=False,
+        blank=False,
+        on_delete=models.CASCADE,
+        related_name="files",
+    )
+
+    objects = DatasetVersionFileQuerySet.as_manager()
+
+    @property
+    def full_uri(self):
+        return f"{self.dataset_version.dataset.id}/{self.dataset_version.id}/{self.uri}"
```

Comment:
Something to consider: maybe storing the full URI would be more robust - to discuss

Reply:
Yes indeed